### PR TITLE
pipeline: hasher bug

### PR DESCRIPTION
**Pull request by acud**
There was a bug in our hashing pipeline that was discovered in #1175.
The problem was that we assumed that the trie depth was known in advanced, however, when wrapping deeper levels under certain circumstances the upper levels would also wrap, resulting in a trie that is deeper than originally assumed. The result was that the trie was being traversed only until the preassumed depth, leaving us with an "empty" level once it has wrapped into the next.

This has been demonstrated with an explicit regression test that is now in the `hashtrie` package. Furthermore, unit tests have now been added to the `hashtrie` package. It covers various scenarios and edge cases.

The `Sum` method logic has now been inlined into the method instead of the helper method, and the logic has been significantly simplified, with extensive comments and explanations on how things work, with the hope that this improves developer onboarding and open-source contributions, which have revealed this bug.

The logic of generating the root hash has changed and indeed simplified. In the past, we would wrongly assume which level we're aiming at and try to read the resulting data out of that level. However as mentioned, this did not take into account the case where several subsequent levels were wrapped as a result of one level wrap (causing the trie to balance, basically).
After the simplification we just iterate on all levels and constantly carry over hashes or wrap levels as needed, carrying over the result to higher level. In the end of the operation, we just take the resulting hash from the top level of the trie.

I have used `TestRegression` to verify that this problem exists on current `master` and is solved with this change, as well as manual tests. In the unit tests added to the `hashtrie` package, I have chosen an approach where one predefined hash is constantly being written to the trie, with constant spans of value `1`. This allows the eventual root chunk span to be compared against the linear number of writes that were done in the test. Also, custom branching and chunk size values were used in tests in order to decrease the number of writes needed in order to trigger a level wrap of the trie.

### CODE REVIEW

File: pkg/storage/mock/storer.go [hunk 1]
```diff
@@ -93,7 +93,9 @@ func (m *MockStorer) Put(ctx context.Context, mode storage.ModePut, chs ...swarm
 			po := swarm.Proximity(ch.Address().Bytes(), m.baseAddress)
 			m.bins[po]++
 		}
-		m.store[ch.Address().String()] = ch.Data()
+		b := make([]byte, len(ch.Data()))
```
Comment (acud) [inline pkg/storage/mock/storer.go:96]: this is needed since the chunk feeder shares memory across calls to the pipeline. this is in order to avoid multiple allocations. this change mimics the behavior of shed and localstore, and copies the data from the call into the in-memory store

**Review by janos — APPROVED** [review]
👏

File: pkg/file/pipeline/hashtrie/hashtrie.go [hunk 2]
```diff
@@ -12,15 +12,21 @@ import (
...
 
-var errInconsistentRefs = errors.New("inconsistent reference lengths in level")
+var (
+	errInconsistentRefs = errors.New("inconsistent references")
+	errTrieFull         = errors.New("trie full")
+)
+
+const maxLevel = 8
 
 type hashTrieWriter struct {
```
Comment (zelig) [inline pkg/file/pipeline/hashtrie/hashtrie.go:22]: call this just writer

File: pkg/file/pipeline/hashtrie/hashtrie.go [hunk 3]
```diff
@@ -12,15 +12,21 @@ import (
 	"github.com/ethersphere/bee/pkg/swarm"
 )
 
-var errInconsistentRefs = errors.New("inconsistent reference lengths in level")
+var (
+	errInconsistentRefs = errors.New("inconsistent references")
```
Comment (zelig) [inline pkg/file/pipeline/hashtrie/hashtrie.go:16]: not  exported?

File: pkg/file/pipeline/hashtrie/hashtrie.go [hunk 4]
```diff
@@ -12,15 +12,21 @@ import (
...
+	errTrieFull         = errors.New("trie full")
```
Comment (zelig) [inline pkg/file/pipeline/hashtrie/hashtrie.go:17]: i thought we agreed this should be handled but  its ok

**Review by zelig — APPROVED** [review]
LGTM. please change trie to tree everywhere. this is not  a trie

**Comment by Eknir** [comment]
This PR is ready for review, but not connected to an issue. The issue which belongs to the PR is still in the needs confirmation pipeline. Please properly manage your issues.

File: pkg/file/pipeline/hashtrie/hashtrie.go (hunk 3 above)
Comment (acud) [reply pkg/file/pipeline/hashtrie/hashtrie.go:16]: what for?

**Comment by acud** [comment]
> change trie to tree everywhere

Let's throttle renaming. I'll do this in a separate PR
